### ISSUE TITLE
Potential Memory Leak in `DefaultDynamoDbEnhancedAsyncClient.batchGetItem(BatchGetItemEnhancedRequest)`

### ISSUE THREAD

**Issue by lorenz-scalable** [aws/aws-sdk-java-v2#4712]
### Describe the bug

We use DynamoDbEnhancedAsyncClient in a Spring Boot 3.1.5 application. It uses Webflux and Reactor for asynchronous processing.

Under high load, we can see a lot of memory allocations that are never released. Those memory allocations are coming from `DefaultDynamoDbEnhancedAsyncClient.batchGetItem(BatchGetItemEnhancedRequest)`. You can see the memory profile in the attached flame graph:
<img width="1072" alt="Screenshot 2023-11-21 at 16 19 08" src="https://github.com/aws/aws-sdk-java-v2/assets/95355196/34f87dc0-1cbf-45c8-966d-a2c1a5f7ae51">

I will also include a small example in the reproduction steps.

### Expected Behavior

The batch requests should release the allocated memory after the requests were processed.

### Current Behavior

The requests allocate memory and don't release it.

### Reproduction Steps

```kotlin
fun findItemsByIds(
    ids: List<String>, 
    dynamoDbClient: DynamoDbEnhancedAsyncClient, 
    itemTable: DynamoDbAsyncTable<Item>
): Mono<List<Item>> {
    val chunkedIds = ids.chunked(MAX_GET_CHUNK_SIZE)

    val batchPublishers = chunkedIds.map { idChunk ->
        dynamoDbClient.batchGetItem(
            BatchGetItemEnhancedRequest.builder()
                .readBatches(
                    idChunk.map { id ->
                        ReadBatch.builder(Item::class.java)
                            .mappedTableResource(itemTable)
                            .addGetItem(
                                Key.builder()
                                    .partitionValue(id)
                                    .build()
                            )
                            .build()
                    }
                )
                .build()
        )
        .toFlux()
        .map { page ->
            page.resultsForTable(itemTable)
        }
    }

    return Flux.merge(batchPublishers)
        .collectList()
        .map { it.flatten() }
}
```

This is more of a pseudo code example, to understand to usage of the API. 

### AWS Java SDK version used

2.21.5

### JDK version used

17

### Operating System and version

Linux Alpine

**Comment by lorenz-scalable** [comment]
Some more context:
<img width="496" alt="Screenshot 2023-11-21 at 18 04 25" src="https://github.com/aws/aws-sdk-java-v2/assets
<img width="497" alt="Screenshot 2023-11-21 at 18 05 08" src="https://github.com/aws/aws-sdk-java-v2/assets/95355196/e773782b-f984-4e3b-aef9-0bb2da9c9069">
/95355196/c5715120-dfe3-4215-a52d-e9d4ae81fce1">

**Comment by debora-ito** [comment]
@lorenz-scalable thank you for reaching out. 

When analyzing memory consumption, flame graphs only show when the allocation happened but not when the memory was deallocated. 

Have you analyzed the heap dump? Can you share it?

**Comment by lorenz-scalable** [comment]
Unfortunately I am not allowed to create or share heap dumps of production systems. I will try to utilise JDK Old Object Sample instead and come back with results.

**Comment by debora-ito** [comment]
Understood. 

So you only experience this memory issue in production? Are you able to reproduce in a non-production environment? In order to root cause it, it would be helpful to have a repro code.

**Comment by lorenz-scalable** [comment]
> Understood.
> 
> So you only experience this memory issue in production? Are you able to reproduce in a non-production environment? In order to root cause it, it would be helpful to have a working repro code (not a pseudo code).

Yes, we only see this on PROD. It is hard for us to reproduce, because we make about 7k requests per second to DynamoDB and reproducing this on a non-PROD account would be very expensive for us.

I had a look at JDK live object samples and the only suspicious part is that the Dax `RetryHandler` (package: `software.amazon.dax.retry`) keeps a lot of memory for `byte[]` arrays.

We can also see a correlation between latency spikes and memory utilization.  There are also some warning logs pointing in the same direction:
```
An exceptionCaught() event was fired, and it reached at the tail of the pipeline. It usually means the last handler in the pipeline did not handle the exception.

Logger Name: io.netty.channel.DefaultChannelPipeline
Stack Trace: io.netty.handler.timeout.ReadTimeoutException: null
Thread Name: dax-NettyEventLoop-0-1
```

**Comment by debora-ito** [comment]
The Dax Client is maintained by the Dax team, not the Java SDK team, so we can't tell in detail how it handles memory. I can raise this issue to the Dax Team, and see if they have any pointers.

Which version of the `amazon-dax-client` are you using? I know they fixed some memory leak issues in version `2.0.2` -
https://search.maven.org/artifact/software.amazon.dax/amazon-dax-client

**Comment by lorenz-scalable** [comment]
We use version `2.0.4` for `software.amazon.dax:amazon-dax-client`

**Comment by lorenz-scalable** [comment]
@debora-ito how can we raise it to the Dax team? Is there another repo for the client or would you raise it internally and keep this issue open for communication?

**Comment by debora-ito** [comment]
I can raise it internally, but if you have access to a AWS Support plan you can open a support case and talk directly to the DAX team. It's ok if you don't have access, just let me know what you prefer.

**Comment by debora-ito** [comment]
@lorenz-scalable is this still relevant? Have you reached out to the Dax team?

**Comment by lorenz-scalable** [comment]
@debora-ito sorry for my late reply. The issue has been resolved. I will reach out to the Dax team directly if it comes back. Thank you for your help :)

**Comment by bhoradc** [comment]
Thanks for the confirmation @lorenz-scalable, closing the issue.
